Feat/let users authorise their sender domain in the plugin - MAILPOET-4302

triple0t opened the pull request:
[MAILPOET-4302](https://mailpoet.atlassian.net/browse/MAILPOET-4302)

This PR provides the ability for a user to add and verify their sender domain from the plugin

It also includes a basic check for DMARC policy and gives a warning if it detects the domain has restrictive DMARC policies.

Testing:
* from the Settings page. test with unauthorized email addresses and with domains having  DMARC policy set to reject or quarantine. e.g @automattic.com or @mailpoet.com 
* From the Newsletter sending page. perform the same test.
* You should be able to authorized an email address, add a new sender domain and verify the sender domain

For QA:
Please retest this PR: https://github.com/mailpoet/mailpoet/pull/4229

Thank you

<img width="890" alt="Screenshot 2022-07-22 at 10 16 27 AM" src="https://user-images.githubusercontent.com/30554163/180426885-d7ac81ea-4a4e-459c-8aa3-d03289abac11.png">

<img width="1619" alt="Screenshot 2022-07-22 at 10 16 14 AM" src="https://user-images.githubusercontent.com/30554163/180426915-b43eb9aa-ef85-4bf4-ac37-c55359abab22.png">

triple0t committed:
Add API methods for fetching, creating and verifying
sender domain

MAILPOET-4302

triple0t committed:
Add DMARC Policy Checker class and
getDomainDmarcPolicy method

MAILPOET-4302

triple0t committed:
Add AuthorizedSenderDomainController class and tests

MAILPOET-4302

triple0t committed:
Add mailpoet_verified_sender_domains to newsletters and settings page

Update tests

MAILPOET-4302

triple0t committed:
Implement Show sender domain warning for settings

This would show the sender domain warning message for
domains that are not verified and have a Retricted DMARC Policy

MAILPOET-4302

triple0t committed:
Add mailpoet_all_sender_domains to Newsletter and settings

MAILPOET-4302

triple0t committed:
Add API methods for creating and verifying sender domains

We also included an API method for fetching Sender Domains

MAILPOET-4302

triple0t committed:
Implement the Manage Sender Domain component and

The AuthorizeSenderDomainModal

MAILPOET-4302

triple0t committed:
Add clean up action after domain is verified

MAILPOET-4302

triple0t committed:
Feat: Add classname to tabs

This would give us the ability to selectively display or hide the
tab component with CSS classes without causing issues
with the JavaScript code

MAILPOET-4302

triple0t committed:
Update SenderDomainModal and AuthorizeSenderEmailModal

Provide the ability to display content without modal

I included the `useModal` option for backward compatibility and
to not break any other components

MAILPOET-4302

triple0t committed:
Add AuthorizeSenderEmailAndDomainModal component

This component wraps the AuthorizeSenderEmailModal  and
AuthorizeSenderDomainModal and displays them in a Tabbed
manner

This gives the user the ability to either authorize an email or verify
a domain

MAILPOET-4302

triple0t committed:
Use unidirectional data store, remove unnecessary useState

MAILPOET-4302

triple0t committed:
Show sender domain verification modal on newsletter
sending page

MAILPOET-4302

triple0t committed:
General refactor

Use a single click handler for css lookup

Set to mailpoet-js-button-authorize-email-and-sender-domain

MAILPOET-4302

triple0t committed:
Update ManageSenderDomain component UI

Add DomainKeyComponent to assist with rendering the input fields
and DomainStatusComponent for rendering the DNS records status

MAILPOET-4302

triple0t committed:
Add tracking for authorized email and sender domain

MAILPOET-4302

triple0t committed:
Update DMARC Status check

Handle edge case for domains where DMARC `p` policy is set to  reject or quarantine but `sp` (subdomain policy) is set to none

The previous implementation will return dmarcStatus === none when
sp is none but will not check for p policy as well.

If the `sp` is reject or quarantine, it would supersede the `p` status

MAILPOET-4302

triple0t committed:
Update information provided for tracking

MAILPOET-4302

costasovo commented on mailpoet/tests/integration/Services/BridgeTest.php:
```diff
@@ -333,6 +333,167 @@ public function testItReturnsAnEmptyArrayIfNoNullForAuthorizedParam() {
+    $api = Stub::make(new API(null), ['getAuthorizedSenderDomains' => $data], $this);
+    $this->bridge->api = $api;
+
+    $result = $this->bridge->getAuthorizedSenderDomains('all');
```
I guess the parameter should be skipped, as the comment suggests.

costasovo commented on mailpoet/tests/integration/Services/BridgeTest.php:
```diff
@@ -333,6 +333,167 @@ public function testItReturnsAnEmptyArrayIfNoNullForAuthorizedParam() {
+      $this->markTestSkipped("Skipping, 'WP_TEST_MAILER_MAILPOET_API' not set.");
+    }
+
+    $api = new API($apiKey, new WPFunctions());
```
I think it would be better to mock API responses instead of calling the live API here. I assume the API endpoints are covered by tests on the bridge side.

costasovo commented on mailpoet/lib/Util/DmarcPolicyChecker.php:
```diff
@@ -0,0 +1,58 @@
+<?php
```
We add `declare(strict_types = 1);` to all new PHP files

costasovo commented on mailpoet/lib/Util/DmarcPolicyChecker.php:
```diff
@@ -0,0 +1,58 @@
+    }
+
+    // policy can either be reject or quarantine or none
+    $dmarcStatus = $dmarcInfo['sp'] ?? $dmarcInfo['p'] ?? self::POLICY_NONE;
```
I would add brackets so that one doesn't have to think about the order when reading the code.

costasovo commented on mailpoet/lib/Services/AuthorizedSenderDomainController.php:
```diff
@@ -0,0 +1,68 @@
+   * returns `true` if domain has Retricted policy e.g policy === reject or quarantine
+   * otherwise returns `false`
+   */
+  public function isDomainDmarcRetricted(string $domain): bool {
```
There is a small typo: `Retricted` > `Restricted`

costasovo commented on mailpoet/lib/AdminPages/Pages/Newsletters.php:
```diff
@@ -196,6 +202,8 @@ public function render() {
 
     $data['authorized_emails'] = $this->bridge->getAuthorizedEmailAddresses();
 
+    $data['verified_sender_domains'] = $this->senderDomainController->getVerifiedSenderDomains();
```
It seems that we always try those even though the user might not use MSS. In such a case we return `null`, but we still call the API without API key. I think we should check here whether the sending method is MSS.

costasovo commented on mailpoet/lib/AdminPages/Pages/Settings.php:
```diff
@@ -97,6 +103,7 @@ public function render() {
       ],
       'built_in_captcha_supported' => $this->captcha->isSupported(),
       'authorized_emails' => $this->bridge->getAuthorizedEmailAddresses(),
+      'verified_sender_domains' => $this->senderDomainController->getVerifiedSenderDomains(),
```
Same as the above. I think we should fetch verified domains only when a is using MSS.

costasovo commented on mailpoet/assets/js/src/settings/pages/basics/default_sender.tsx:
```diff
@@ -22,6 +37,36 @@ export function DefaultSender() {
   };
+
+  const checkSenderEmailDomain = (email: string) => {
+    const emailAddressDomain = email.split('@').pop().toLowerCase();
```
I think we need to skip this check when sending method is not MSS.

costasovo commented on mailpoet/lib/API/JSON/v1/Settings.php:
```diff
@@ -261,13 +261,93 @@ public function checkDomainDmarcPolicy($data = []) {
+      $response = $this->senderDomainController->createAuthorizedSenderDomain($domain);
+    } catch (\InvalidArgumentException $e) {
+      if (
+        $e->getMessage() === AuthorizedSenderDomainController::AUTHORIZED_SENDER_DOMAIN_ERROR_ALREADY_CREATED
```
I think it would be nicer to use exception codes for the logic instead of messages, but it is not a big deal

costasovo commented on mailpoet/assets/js/src/common/manage_sender_domain/manage_sender_domain.tsx:
```diff
@@ -0,0 +1,100 @@
+  );
+}
+
+ManageSenderDomain.propTypes = {
+  max_width: PropTypes.string,
+  rows: PropTypes.arrayOf(
+    PropTypes.shape({
+      domain: PropTypes.string.isRequired,
+      dns: PropTypes.arrayOf(
+        PropTypes.shape({
+          host: PropTypes.string,
+          value: PropTypes.string,
+          type: PropTypes.string,
+          status: PropTypes.string,
+          message: PropTypes.string,
+        }),
+      ).isRequired,
+    }),
+  ).isRequired,
+  verifyDnsButtonClicked: PropTypes.func,
+};
```
For other .tsx components, we only declare types for Props. I guess it is enough. I need to check it though.

costasovo replied:
I checked the component, and it seems that without the prop type checks don't trigger in runtime. So I'll keep the prototypes in this one.

JanJakes replied:
Why do you need to check them at runtime? Do we catch those failures? If not, then I'm not sure if it's useful. Unknown data should be validated and sanitized somewhere, and then TS just can statically check if the rest is good. Duplicating it in PropTypes to make it crash sooner seems unnecessary to me.

costasovo replied:
I'm not sure. Maybe it was handy for the author when developing the API and component at the same time. I'm just guessing. I will remove it.

JanJakes replied:
Ahh, you're finishing this PR, now I get it, you don't know. I guess the only reason may be not everybody knows about this decision to use TS types only for props.

costasovo reviewed (changes requested):
Wow! What a huge amount of work! I've found some small issues that need addressing.

costasovo committed:
Add declare(strict_types = 1); to newly added PHP files

[MAILPOET-4302]

costasovo committed:
Fix typo in isDomainDmarcRestricted method name

[MAILPOET-4302]

costasovo committed:
Refactor sender domain related bridge integration tests to use mocks

[MAILPOET-4302]

costasovo committed:
Refactor AuthorizedSenderDomainControllerTest to avoid live api calls

[MAILPOET-4302]

costasovo committed:
Fetch authorized domains and emails for JS Apps only when MSS is active

This prevents calling bridge API in cases when user use different sending methods.
[MAILPOET-4302]

costasovo committed:
Refactor sender address errors rendering on send page

Both auht. email error and DMARC check warning were rendered via parsley error.
There was not enough space for both messages to display and also both were rendered
as errors that block sending. But so far we allow sending when the DMARC warning is present.

This commit changes displaying of both error messages so that they are rendered in
a custom component and use the parsley only to mark the input red.
[MAILPOET-4302]

costasovo committed:
Prevent triggering domain and authorized email checks when MSS is not active

[MAILPOET-4302]

costasovo committed:
Skip DMARC address check when email domain is freemail

[MAILPOET-4302]

costasovo committed:
Removing protypes from Typescript component

We use only typed props in components written in TS.
[MAILPOER-4302]

costasovo committed:
Remove unused max_width property from manage_sender_domain

[MAILPOET-4302]

costasovo committed:
Use common/button component fot copy button

The proprietary button had displaying issues. This commit add props
that enable connection with a tooltip into the common button and replaces
the proprietary button with the common component.
[MAILPOET-4302]

costasovo committed:
Display correct tab in sender domain/email configuration modal based on clicked link

Previously when both errors were present authorized email tab was always opened no matter what error was clicked.
[MAILPOET-4302]

costasovo committed:
Unify loaders size in sender domain/email tabs and render in column

This makes sure that the modal has some reasonable initial size (small column).
[MAILPOET-4302]

costasovo committed:
Update list of popular free ESP domains

[MAILPOET-4302]

costasovo committed:
Hide sender domains error when editing sender address

[MAILPOET-4302]

costasovo committed:
Add basic acceptance tests for sender domain verification

These tests test that basic checks works in both settings and send pages.
We can't verify the test domain for acceptance test docker so the tests just check the error state.
[MAILPOET-4302]

costasovo commented:
@lysyjan, I've fixed the issues I found during the code review. Well, except [this one](https://github.com/mailpoet/mailpoet/pull/4265#discussion_r929734378) as I considered just a minor issue and I found more things that needed some changes.

It might be worth checking the whole PR.

I've also fixed a couple of issues I found when working on the fixes.

### 1) I've improved the appearance of the errors on the send page
**Before**
The error would probably not fit in case translations were a bit longer.

<img width="387" alt="Screenshot 2022-08-02 at 10 06 25" src="https://user-images.githubusercontent.com/1082140/182324890-81429522-bce7-4417-a2dd-8b82d4ad9cc0.png">


**After**

<img width="391" alt="Screenshot 2022-08-02 at 10 07 28" src="https://user-images.githubusercontent.com/1082140/182325075-d84842a8-2333-4957-b741-48d95ecdd9d8.png">


### 2) When both errors were displayed, the link from the sender domain error was opening the wrong tab in modal
Now when a user clicks on the link in the authorized email error, the modal opens with the preselected Authorized Emails tab, and when a user clicks on the link in the sender domain error, it opens the sender domain settings tab

### 3) I added skipping checks of domain sender settings for free email addresses
MSS allows using free emails as senders. In those cases, a user can't change DNS (e.g. gmail.com), so we will not display the error.

### 4) A few minor UX/UI issues
* fixed initial width of modal by displaying the initial spinner in columns and unified size of the spinner
* fixed the appearance of copy to clipboard icons in the domain settings table
* added hiding of the sender error when user edits the sender address (we have the same behavior for the auth. email error)

### 5) I added a couple of basic acceptance tests for the new functionality

lysyjan commented on mailpoet/tests/integration/Util/DmarcPolicyCheckerTest.php:
```diff
@@ -0,0 +1,55 @@
+  public function testItReturnsNoneWhenDomainDoesNotHaveDmarc() {
+    $domain = 'example.com'; // has no TXT records
+    $result = $this->dmarcPolicyChecker->getDomainDmarcPolicy($domain);
+    expect($result)->equals('none');
```
I think that we could use constants from the DmarcPolicyChecker class in those comparisons.

lysyjan commented on mailpoet/lib/Util/DmarcPolicyChecker.php:
```diff
@@ -0,0 +1,64 @@
+   */
+  public function getDomainDmarcPolicy(string $domain): string {
+
+    if (!$domain) {
```
I noticed that this case is not covered in the test, but it is nothing urgent.

lysyjan commented on mailpoet/lib/Services/Bridge.php:
```diff
@@ -139,6 +139,57 @@ public function createAuthorizedEmailAddress(string $emailAdress) {
+   */
+  public function createAuthorizedSenderDomain(string $domain): array {
+    $data = $this
+    ->getApi($this->settings->get(self::API_KEY_SETTING_NAME))
```
It seems that the indentation here and in the following method is incorrect.

lysyjan commented on mailpoet/lib/Services/Bridge.php:
```diff
@@ -139,6 +139,57 @@ public function createAuthorizedEmailAddress(string $emailAdress) {
+
+  /**
+   * Create a new Sender domain record
+   * returns an Array of DNS response or array of error
```
I have no idea how the response can look. An example or a link to the bridge documentation in the comment could be great here because our codebase and team are growing fast. This can be applicated to the other methods too.

lysyjan commented on mailpoet/lib/Services/Bridge/API.php:
```diff
@@ -215,6 +217,89 @@ public function createAuthorizedEmailAddress(string $emailAddress): array {
+   * Get a list of sender domains
+   * Fetched from API
+   */
+  public function getAuthorizedSenderDomains(): ?array {
```
This situation with examples is similar to the Bridge class.

lysyjan commented on mailpoet/lib/Services/Bridge/API.php:
```diff
@@ -215,6 +217,89 @@ public function createAuthorizedEmailAddress(string $emailAddress): array {
+      ];
+      $this->loggerFactory->getLogger(LoggerFactory::TOPIC_BRIDGE)->error('verifyAuthorizedSenderDomain API call failed.', $logData);
+
+      $fallbackError = sprintf($this->wp->__('An error has happened while performing a request, the server has responded with response code %d'), $code);
```
If I am not wrong, we use the translate function directly and not via our functions wrapper, right? Also, the second parameter with the domain is missing.

lysyjan commented on mailpoet/lib/Services/Bridge/API.php:
```diff
@@ -215,6 +217,89 @@ public function createAuthorizedEmailAddress(string $emailAddress): array {
+      return ['error' => $errorData, 'status' => false];
+    }
+
+    return is_array($responseBody) ? $responseBody : [];
```
Is it a correct state when responseBody is not an array? Should we log it? TBH, I don't know, so I want to ask only.

costasovo replied:
I'll add some logging

lysyjan commented on mailpoet/tests/integration/Services/BridgeTest.php:
```diff
@@ -333,6 +333,167 @@ public function testItReturnsAnEmptyArrayIfNoNullForAuthorizedParam() {
+  }
+
+  public function testItReturnsSenderDomainsDnsRecords() {
+    $data = array (
```
Wau, longer array syntax, I haven't seen it for ages in the new code :)

I tried to check the current state after fixes, and this array is duplicated in the code. I guess we can reuse the same array from the mock. Is it possible?

lysyjan commented on mailpoet/lib/API/JSON/v1/Settings.php:
```diff
@@ -246,6 +252,102 @@ public function confirmSenderEmailAddressIsAuthorized($data = []) {
+
+    $domain = strtolower(trim($domain));
+
+    $response = ['isDmarcPolicyRetricted' => $this->senderDomainController->isDomainDmarcRestricted($domain)];
```
Here is another typo in the array key `isDmarcPolicyRetricted`.

lysyjan commented on mailpoet/lib/API/JSON/v1/Settings.php:
```diff
@@ -246,6 +252,22 @@ public function confirmSenderEmailAddressIsAuthorized($data = []) {
+
+    $domain = trim($domain);
+
+    $response = ['isDmarcPolicyRetricted' => $this->senderDomainController->isDomainDmarcRetricted($domain)];
```
Here is another typo in the array key `isDmarcPolicyRetricted`.

lysyjan commented on mailpoet/lib/API/JSON/v1/Settings.php:
```diff
@@ -246,6 +252,22 @@ public function confirmSenderEmailAddressIsAuthorized($data = []) {
+
+    if (!$domain) {
+      return $this->badRequest([
+        APIError::BAD_REQUEST => WPFunctions::get()->__('No sender domain specified.', 'mailpoet'),
```
We should use the translate function directly. Please, check the whole class here.

lysyjan commented on mailpoet/tests/integration/Services/AuthorizedSenderDomainControllerTest.php:
```diff
@@ -18,41 +19,62 @@ class AuthorizedSenderDomainControllerTest extends \MailPoetTest {
   }
 
+  public function testItReturnsEmptyArrayWhenNoVerifiedSenderDomains() {
+    $expectaton = Expected::once([]); // with empty array
```
I guess this variable should name `expectation` instead of `expectaton`.

lysyjan commented on mailpoet/tests/integration/Services/AuthorizedSenderDomainControllerTest.php:
```diff
@@ -65,8 +87,20 @@ public function testItReturnsFalseWhenDmarcIsNotEnabled() {
-  private function getController(): AuthorizedSenderDomainController {
+  public function testItReturnsDmarcStatus() {
+    $controller = $this->getController();
+    $isRetricted = $controller->getDmarcPolicyForDomain('example.com');
```
Another typo is here in the restricted word.

lysyjan commented on mailpoet/lib/Services/AuthorizedSenderDomainController.php:
```diff
@@ -23,37 +29,101 @@ public function __construct(
+  /**
+   * Get the most recent cached record of Bridge::getAuthorizedSenderDomains
+   */
+  public function getDomainRecords($domain = ''): array {
```
We could use a type for the domain argument here.

lysyjan commented on mailpoet/lib/Services/AuthorizedSenderDomainController.php:
```diff
@@ -23,37 +29,101 @@ public function __construct(
-    $records = $this->bridge->getAuthorizedSenderDomains();
-    $domains = array_keys($records);
-    return $domains;
+    $records = $this->currentRecords = $this->bridge->getAuthorizedSenderDomains();
```
Is this correct? Do we override the `currentRecords` properly on each call? If yes, we can simplify this method.
The same logic is in the `getVerifiedSenderDomains` method.

costasovo replied:
Yep. This is done because of caching. `$this->currentRecords` is also used in `AuthorizedSenderDomainController::getDomainRecords`. But with this implementation, the method `getDomainRecords` will not work properly if the cache is not prepopulated. I'll refactor it a bit.

lysyjan commented on mailpoet/tests/integration/Services/AuthorizedSenderDomainControllerTest.php:
```diff
@@ -75,6 +76,75 @@ public function testItReturnsEmptyArrayWhenNoVerifiedSenderDomains() {
+    $this->expectExceptionMessage('Sender domain already verified');
+
+    $domains = ['testdomain.com' => [
+      ['status' => 'valid'],
```
It's better to use constant here to avoid typos. But only for the next time.

lysyjan commented on mailpoet/tests/integration/Services/AuthorizedSenderDomainControllerTest.php:
```diff
@@ -75,6 +76,75 @@ public function testItReturnsEmptyArrayWhenNoVerifiedSenderDomains() {
+    $verifySenderDomainsExpectaton = Expected::once(['error' => $errorMessage, 'status' => false]);
+
+    $bridgeMock = $this->make(Bridge::class, [
+      'getAuthorizedSenderDomains' => $getSenderDomainsExpectaton,
```
Another typo in the word `Expectaton`.

lysyjan commented on mailpoet/assets/js/src/common/check_sender_domain_dmarc_policy.ts:
```diff
@@ -0,0 +1,54 @@
+
+  try {
+    const res = await makeApiRequest(emailAddressDomain);
+    const isDmarcPolicyRetricted = Boolean(res?.data?.isDmarcPolicyRetricted);
```
Another typo

lysyjan reviewed (commented):
Great job with those fixes and improvements!
I have noticed several details. It's up to you if you want to address them. I'm fine with moving this PR forward.

costasovo committed:
Use constants and add missing test case in DmarcPolicyCheckerTest

[MAILPOET-4302]

costasovo committed:
Fix indentation in Bridge

[MAILPOET-4302]

costasovo committed:
Add links to sender domains API docs

[MAILPOET-4302]

costasovo committed:
Use WordPress translation functions directly

[MAILPOET-4302]

costasovo committed:
Reuse mocked api data in Bridre test case

[MAILPOET-4302]

costasovo committed:
Add logging of unexpected formats from sender domain endpoints on Bridge API

[MAILPOET-4302]

costasovo committed:
Fix variable names and comments typos

[MAILPOET-4302]

costasovo committed:
Refactor response caching in AuthorizedSenderDomainController

In the previous implementation the getDomainRecords couldn't be used without
calling getAllSenderDomains or getVerifiedSenderDomains upfront.
[MAILPOET-4302]

costasovo commented:
@lysyjan I addressed all your remarks [except this one optional](https://github.com/mailpoet/mailpoet/pull/4265#discussion_r937508951).

veljkho reviewed (changes requested):
Nice try!
When I enter an unauthorized sender domain, I just can't get the message "Email violates Sender Domain’s DMARC policy. Please set up sender authentication.". This happens only once when using moonbridge.com domain (and I have moonbridge.rs). If I try to use any other domain... I just can't get the possibility to authorize it (even the tab is missing in the popup).

Check this GIF: https://d.pr/i/EGgGqp
My authorized domain is: moonbridge.rs ✅ 
Using moonbridge.com I could get the proper response ✅  (for some reason it works with this domain, it's just a tld change)
Using any other domain, literally anything like moon.com, sdfds.com etc.. no response at all ❌

costasovo commented:
@veljkho Thanks for checking!

> Using any other domain, literally anything like moon.com, sdfds.com etc.. no response at all ❌

I tested the feature with your examples, and actually, I think this behavior is correct. The ticket is about displaying **a specific warning for domains that have set restrictive DMARC policies**.  In case the domain doesn't have set the restrictive policy, we do not display the warning (see point 4 on [the ticket](https://mailpoet.atlassian.net/browse/MAILPOET-4302)). In your example, the domain `moonbridge.com` has set restrictive DMARC policy (`v=DMARC1;p=quarantine;sp=quarantine;adkim=r;aspf=r;pct=100;fo=0;rf=afrf;ri=86400`) but for domains `moon.com` or `moonbridge.bb` we can't find any DMARC policies at all so that means we treat the the policy as non-restrictive and we don't display the warning.

costasovo committed:
Fix errors caught by translations code sniff

[MAILPOET-4302]

veljkho reviewed (approved):
Verified. I will retest this on release build.
